module/VuFind/src/VuFind/Controller/Plugin/DbUpgrade.php

```diff
@@ -185,17 +185,33 @@ protected function getCharsetAndCollationProblemsForTableColumns(
         $table,
         $collation
     ) {
+        $collation = strtolower($collation);
+
         // Get column summary:
         $sql = "SHOW FULL COLUMNS FROM `{$table}`";
         $results = $this->getAdapter()->query($sql, DbAdapter::QUERY_MODE_EXECUTE);
 
+        // Get expected column types:
+        // Parse column names out of the CREATE TABLE SQL, which will always be
+        // the first entry in the array; we assume the standard mysqldump
+        // formatting is used here.
+        preg_match_all(
+            '/^  `([^`]*)`\s+([^\s,]+)[\t ,]+.*$/m',
+            $this->dbCommands[$table][0],
+            $matches
+        );
+        $expectedTypes = $matches[2];
+
         // Load details:
         $retVal = [];
-        foreach ($results as $current) {
-            if (!empty($current->Collation)
-                && strtolower($current->Collation) !== strtolower($collation)
-            ) {
-                $retVal[$current->Field] = (array)$current;
+        foreach ($results as $i => $current) {
+            if (!empty($current->Collation)) {
+                $normalizedCollation = strtolower($current->Collation);
+                if ($normalizedCollation !== $collation
+                    && $expectedTypes[$i] !== 'json'
```

Comment:
I'm not sure I understand why we completely ignore the collation check for json. Shouldn't it get the correct collation for the table, regardless of other details?

If we really do never want to touch collation on json fields, I wonder if we should move the expectedTypes check earlier, since I don't think there's any reason to normalize and compare collations if we're going to skip them anyway.

Reply:
No, collation is always utf8mb4_bin for json fields unless something else is specified for the column, regardless of the table collation. I don't think there's a reason to change that, but this can be revisited if necessary.

I moved the check earlier and changed to an associative array to ensure that any discrepancy in column order or extra columns don't break the check.